src/lib.rs

```diff
@@ -105,42 +97,200 @@ impl Default for FxHasher {
 impl FxHasher {
     #[inline]
     fn add_to_hash(&mut self, i: usize) {
-        self.hash = self.hash.rotate_left(5).bitxor(i).wrapping_mul(K);
+        self.hash = self.hash.wrapping_add(i).wrapping_mul(K);
     }
 }
 
 impl Hasher for FxHasher {
     #[inline]
-    fn write(&mut self, mut bytes: &[u8]) {
-        // Ensure all bytes will be consumed
-        const _: () = assert!(size_of::<usize>() <= size_of::<u64>());
-        // Ensure no bytes are discarded by casting to usize
-        const _: () = assert!(size_of::<u32>() <= size_of::<usize>());
-        // Copy the 1 word sized state to a local variable to ensure it
-        // is kept in a register.
-        // See: https://github.com/rust-lang/rustc-hash/pull/34
-        let mut state = self.clone();
-        while let Some(&usize_bytes) = take_first_chunk(&mut bytes) {
-            state.add_to_hash(usize::from_ne_bytes(usize_bytes));
-        }
-        if let Some(&u32_bytes) = take_first_chunk(&mut bytes) {
-            state.add_to_hash(u32::from_ne_bytes(u32_bytes) as usize);
-        }
-        if let Some(&u16_bytes) = take_first_chunk(&mut bytes) {
-            state.add_to_hash(u16::from_ne_bytes(u16_bytes) as usize);
-        }
-        if let Some(&[u8_byte]) = take_first_chunk(&mut bytes) {
-            state.add_to_hash(u8_byte as usize);
-        }
-        *self = state;
+    fn write(&mut self, bytes: &[u8]) {
+        self.write_u64(hash_bytes(bytes));
+    }
+
+    #[inline]
+    fn write_u8(&mut self, i: u8) {
+        self.add_to_hash(i as usize);
+    }
+
+    #[inline]
+    fn write_u16(&mut self, i: u16) {
+        self.add_to_hash(i as usize);
+    }
+
+    #[inline]
+    fn write_u32(&mut self, i: u32) {
+        self.add_to_hash(i as usize);
+    }
+
+    #[inline]
+    fn write_u64(&mut self, i: u64) {
+        self.add_to_hash(i as usize);
+        #[cfg(target_pointer_width = "32")]
+        self.add_to_hash((i >> 32) as usize);
+    }
+
+    #[inline]
+    fn write_u128(&mut self, i: u128) {
+        self.add_to_hash(i as usize);
+        #[cfg(target_pointer_width = "32")]
+        self.add_to_hash((i >> 32) as usize);
+        self.add_to_hash((i >> 64) as usize);
+        #[cfg(target_pointer_width = "32")]
+        self.add_to_hash((i >> 96) as usize);
+    }
+
+    #[inline]
+    fn write_usize(&mut self, i: usize) {
+        self.add_to_hash(i);
+    }
+
+    #[cfg(feature = "nightly")]
+    #[inline]
+    fn write_length_prefix(&mut self, _len: usize) {
+        // Most cases will specialize hash_slice anyway which calls write(),
+        // which encodes the length already.
+    }
```

Comment:
This comment seems orthogonal... it may be true that hash_slice calls write, but for the cases that don't use hash_slice and do call write_length_prefix, shouldn't we hash the length?

Reply:
> This comment seems orthogonal... it may be true that hash_slice calls write, but for the cases that don't use hash_slice and do call write_length_prefix, shouldn't we hash the length?

To be entirely correct, yes. It's just not worth the cost.

Reply:
Your stats above suggest that string hashing accounts for less than 1% of cases, if I understand correctly. So is the extra cost meaningful?

Reply:
Note that the trait also mentions this:


> #### Note to Implementers
> If you’ve decided that your Hasher is willing to be susceptible to Hash-DoS attacks, then you might consider skipping hashing some or all of the len provided in the name of increased performance.

Reply:
One more brief comment? :)

Reply:
I don't mind adding a comment.

> Your stats above suggest that string hashing accounts for less than 1% of cases, if I understand correctly. 

No, it's separated out by string length. In what I posted above it's already ~2% and there's more in a long tail I didn't post (I cut it off at 99%).

Also note that we *do* handle the length correctly for strings and slices, because those go through `write` (the latter through specialization). The only thing I can think of that's reasonably common that needs prefix-freedom in `Hash` through `write_length_prefix` is `VecDeque` since that can't delegate to a slice hash.